--- 
title: "Simulation for Microbiome Analysis"
author: "Kris Sankaran, Saritha Kodikara, and Kim-Anh LÃª Cao"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
# url: your book url like https://bookdown.org/yihui/bookdown
# cover-image: path to the social sharing image like images/cover.jpg
description: |
  Examples accompanying our simulation for microbiome analysis review article, available at: 
link-citations: yes
github-repo: rstudio/bookdown-demo
---

This website accompanies the review, 
[``Leveraging Simulation in Microbiome Data Analysis''](). All examples
discussed in the case studies there can be reproduced by running the code below,
and we have also included additional introductory material about some packages
that simplify our analysis.

## Setup

The `MIGsim` package below provides wrapper functions that we will use
throughout these book. You can install it by running:

```{r, eval = FALSE}
devtools::install_github("krisrs1128/intro-to-simulation/MIGsim")
```

This block loads all packages that will be used in this book.

```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(dplyr)
  library(gamboostLSS)
  library(ggdist)
  library(ggplot2)
  library(glue)
  library(mixOmics)
  library(purrr)
  library(splines)
  library(MIGsim)
  library(scDesigner)
})
theme_set(theme_classic())
```

## Using `SummarizedExperiment`

`scDesigner`'s interface is built with `SummarizedExperiment` in mind. These
data structures simplify manipulation of sequencing experiments. For example,
they distinguish between molecule counts, which are stored in the `assay` slot,
and sample descriptors, which are stored in `colData`. At the same time, these
separate components are nicely synchronized. For example, subsetting samples
from one of these tables automatically subsets the other.

The line below loads a small subset of genera from the
[Atlas](https://microbiome.github.io/tutorials/Atlas.html) experiment, which
profiled the gut microbiomes from 1006 healthy adults in Western Europe.

```{r}
data(atlas)
table(rowData(atlas)$Phylum)
mean(atlas$age)


ix <- colData(atlas)$bmi_group == "obese"
abundances <- assay(atlas)
rowMeans(abundances[, ix])
```

**Exercise**: To practice working with `SummarizedExperiment`
objects, try answering:

* How many genera are available in this experiment object?
* What was the most common phylum in this dataset?
* What was the average participant age?
* What was the average abundance of `Allistipes et rel.` among people in the `obese` BMI group?

_Hint: The most important functions are `assay()`, `rowData()`, and `colData()`._

**Solution**

```{r}
nrow(atlas)
table(rowData(atlas)$Phylum)
mean(atlas$age)

atlas[, atlas$bmi_group == "obese"] |>
  assay() |>
  rowMeans()
```

## Warm-up: A Gaussian Example

Here's a toy dataset to illustrate the main idea of GAMLSS. Each panel in the
plot below represents a different feature (e.g., taxon, gene, metabolite...).
The abundance varies smoothly over time, and in the first three panels, the
trends differ by group assignment.

```{r}
data(exper_ts)
exper_lineplot(exper_ts)
```

We can try to approximate these data with a new simulator. The `setup_simulator`
command takes the template `SummarizedExperiment` object as its first argument.
The second gives an R formula syntax-style specification of GAMLSS parameters
(mean and SD, in this case) dependence on sample properties. The last argument
gives the type of model to fit, in this case, a Gaussian location-shape-scale
model.

```{r}
sim <- setup_simulator(exper_ts, ~ ns(time, df = 7) * group, ~ GaussianLSS()) |>
  estimate(nu = 0.01, mstop = 1000)

sample(sim) |>
  exper_lineplot()
```

**Exercise**: Right now, each panel allows for an interaction between the trend and
group type. Can you define a simulator where the groups have no effect on the 
trends for the first two panels? This is the basis for defining synthetic
negative controls.

```{r, eval = FALSE}
sim <- sim |>
  scDesigner::mutate(
    1:2,
     link = ~ ns(time, df = 7)
  ) |>
  estimate(nu = 0.01, mstop = 1000)

sample(sim) |>
  exper_lineplot()
```

**Solution**: We can modify the formula so that it no longer has an interaction
with group. We just need to remove the `* group` from the original formula in our updated
link function. To ensure that this only applies to the first two panels, we use
1:2 in the first argument of `mutate`. This first argument specifies which 
features to apply the new formula to.
```{r}
sim <- sim |>
  mutate(1:2, link = ~ ns(time, df = 7)) |>
  estimate(nu = 0.01, mstop = 1000)

sample(sim) |>
  exper_lineplot()
```

```{r}
sessionInfo()
```

```{r include=FALSE}
# automatically create a bib database for R packages
knitr::write_bib(c(
  .packages(), 'bookdown', 'knitr', 'rmarkdown'
), 'packages.bib')
```
