<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="4.1 Estimation | Simulation for Microbiome Analysis" />
<meta property="og:type" content="book" />

<meta property="og:description" content="Examples illustrating the use of simulation for microbiome analysis, companion to the review article:" />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Kris Sankaran, Saritha Kodikara, Jingyi Jessica Li, and Kim-Anh Lê Cao" />

<meta name="date" content="2024-09-08" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Examples illustrating the use of simulation for microbiome analysis, companion to the review article:">

<title>4.1 Estimation | Simulation for Microbiome Analysis</title>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #2a211c;
    color: #bdae9d;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #bdae9d;  padding-left: 4px; }
div.sourceCode
  { color: #bdae9d; background-color: #2a211c; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffff00; } /* Alert */
code span.an { color: #0066ff; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #44aa43; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #43a8ed; font-weight: bold; } /* ControlFlow */
code span.ch { color: #049b0a; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #0066ff; font-weight: bold; font-style: italic; } /* Comment */
code span.do { color: #0066ff; font-style: italic; } /* Documentation */
code span.dt { text-decoration: underline; } /* DataType */
code span.dv { color: #44aa43; } /* DecVal */
code span.er { color: #ffff00; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #44aa43; } /* Float */
code span.fu { color: #ff9358; font-weight: bold; } /* Function */
code span.im { } /* Import */
code span.in { color: #0066ff; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #43a8ed; font-weight: bold; } /* Keyword */
code span.op { } /* Operator */
code span.pp { font-weight: bold; } /* Preprocessor */
code span.sc { color: #049b0a; } /* SpecialChar */
code span.ss { color: #049b0a; } /* SpecialString */
code span.st { color: #049b0a; } /* String */
code span.va { } /* Variable */
code span.vs { color: #049b0a; } /* VerbatimString */
code span.wa { color: #ffff00; font-weight: bold; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if bootstrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<!--bookdown:toc:end-->
<!--bookdown:toc:start-->
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="estimation-1" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Estimation</h2>
<p>Let’s use simulation to benchmark network estimation methods using data from
rounds 1 and 2 of the <a href="https://journals.asm.org/doi/10.1128/msystems.00031-18">American Gut
Project</a>. We will
simulate data with known correlation structure and taxa-level marginals
estimated from the study data. The block below reads in the data.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="4.1-estimation-1.html#cb60-1" tabindex="-1"></a><span class="fu">data</span>(amgut)</span>
<span id="cb60-2"><a href="4.1-estimation-1.html#cb60-2" tabindex="-1"></a>amgut</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 45 261 
## metadata(0):
## assays(1): counts
## rownames(45): 326792 181016 ... 364563 301645
## rowData names(0):
## colnames(261): 000001879.1076223 000002437.1076448 ... 000002656.1076186 000001582.1076447
## colData names(167): X.SampleID BarcodeSequence ... Description sequencing_depth</code></pre>
<p>We’ve estimated a zero-inflated negative binomial location-shape-scale
(<code>ZINBLSS</code>) models for each taxon, using a gaussian copula to capture
dependence. We have used the regression formula <code>~log(sequencing_depth) + BMI</code>.
The data structure below captures all the simulator components, and we can swap
pieces in and out to modify the form of the simulator. For example, if we
wanted, we could <code>mutate</code> the family and link function associated with
particular features.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="4.1-estimation-1.html#cb62-1" tabindex="-1"></a>sim <span class="ot">&lt;-</span> <span class="fu">setup_simulator</span>(</span>
<span id="cb62-2"><a href="4.1-estimation-1.html#cb62-2" tabindex="-1"></a>  amgut,</span>
<span id="cb62-3"><a href="4.1-estimation-1.html#cb62-3" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">log</span>(sequencing_depth) <span class="sc">+</span> BMI,</span>
<span id="cb62-4"><a href="4.1-estimation-1.html#cb62-4" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">ZINBLSS</span>()</span>
<span id="cb62-5"><a href="4.1-estimation-1.html#cb62-5" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb62-6"><a href="4.1-estimation-1.html#cb62-6" tabindex="-1"></a>  <span class="fu">estimate</span>(<span class="at">mstop =</span> <span class="dv">100</span>)</span>
<span id="cb62-7"><a href="4.1-estimation-1.html#cb62-7" tabindex="-1"></a>sim</span></code></pre></div>
<pre><code>## [Marginals]
## Plan:
## # A tibble: 6 × 3
##     feature              family                         link
##   &lt;gene_id&gt;             &lt;distn&gt;                       &lt;link&gt;
## 1    326792 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 2    181016 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 3    191687 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 4    326977 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 5    194648 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 6    541301 ZINBI [mu,sigma,nu] ~log(sequencing_depth) + BMI
## 
## Estimates:
## # A tibble: 3 × 2
##   feature fit       
##   &lt;chr&gt;   &lt;list&gt;    
## 1 326792  &lt;glmbsLSS&gt;
## 2 181016  &lt;glmbsLSS&gt;
## 3 191687  &lt;glmbsLSS&gt;
## ... and 42 additional features.
## 
## [Dependence]
## 1 normalCopula with 45 features
## 
## [Template Data]
## class: SummarizedExperiment 
## dim: 45 261 
## metadata(0):
## assays(1): counts
## rownames(45): 326792 181016 ... 364563 301645
## rowData names(0):
## colnames(261): 000001879.1076223 000002437.1076448 ... 000002656.1076186 000001582.1076447
## colData names(167): X.SampleID BarcodeSequence ... Description sequencing_depth</code></pre>
<p>The simulated data is always a <code>SummarizedExperiment</code>. This means that any
workflow that applied to the original data can be applied to the simulated one
without any changes. Notice also that <code>sample</code> defaults to drawing samples from
the same design as the original input experiment (we’ll modify this using the
<code>new_data</code> argument in a minute).</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="4.1-estimation-1.html#cb64-1" tabindex="-1"></a>simulated <span class="ot">&lt;-</span> <span class="fu">sample</span>(sim)</span>
<span id="cb64-2"><a href="4.1-estimation-1.html#cb64-2" tabindex="-1"></a>simulated</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 45 261 
## metadata(0):
## assays(1): counts_1
## rownames(45): 326792 181016 ... 364563 301645
## rowData names(0):
## colnames(261): 000001879.1076223 000002437.1076448 ... 000002656.1076186 000001582.1076447
## colData names(167): X.SampleID BarcodeSequence ... Description sequencing_depth</code></pre>
</div>
<p style="text-align: center;">
<a href="4-microbiome-networks.html"><button class="btn btn-default">Previous</button></a>
<a href="4.2-evaluation-1.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
