<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Multivariate Power Analysis | Simulation for Microbiome Analysis</title>
<meta name="author" content="Kris Sankaran, Saritha Kodikara, Jingyi Jessica Li, and Kim-Anh Lê Cao">
<meta name="description" content="How can we choose sample sizes in more complex bioinformatic workflows, where we simultaneously analyze many features (taxa, genes, metabolites) in concert? While traditional, analytical power...">
<meta name="generator" content="bookdown 0.40 with bs4_book()">
<meta property="og:title" content="Chapter 3 Multivariate Power Analysis | Simulation for Microbiome Analysis">
<meta property="og:type" content="book">
<meta property="og:url" content="https://krisrs1128.github.io/microbiome-simulation/multivariate-power-analysis.html">
<meta property="og:description" content="How can we choose sample sizes in more complex bioinformatic workflows, where we simultaneously analyze many features (taxa, genes, metabolites) in concert? While traditional, analytical power...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Multivariate Power Analysis | Simulation for Microbiome Analysis">
<meta name="twitter:description" content="How can we choose sample sizes in more complex bioinformatic workflows, where we simultaneously analyze many features (taxa, genes, metabolites) in concert? While traditional, analytical power...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Simulation for Microbiome Analysis</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="simulating-differential-abundance.html"><span class="header-section-number">2</span> Simulating Differential Abundance</a></li>
<li><a class="active" href="multivariate-power-analysis.html"><span class="header-section-number">3</span> Multivariate Power Analysis</a></li>
<li><a class="" href="microbiome-networks.html"><span class="header-section-number">4</span> Microbiome Networks</a></li>
<li><a class="" href="batch-effect-correction.html"><span class="header-section-number">5</span> Batch Effect Correction</a></li>
<li><a class="" href="vertical-integration.html"><span class="header-section-number">6</span> Vertical Integration</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="multivariate-power-analysis" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Multivariate Power Analysis<a class="anchor" aria-label="anchor" href="#multivariate-power-analysis"><i class="fas fa-link"></i></a>
</h1>
<p>How can we choose sample sizes in more complex bioinformatic workflows, where we
simultaneously analyze many features (taxa, genes, metabolites) in concert?
While traditional, analytical power analysis often breaks down, simulation can
still be effective. We’ll look at a concrete case study where we try to choose a
good sample size for a sparse partial least squares discriminant analysis
(sPLS-DA) of the Type I Diabetes (T1D) gut microbiome.</p>
<p>First, we’ll load the required packages. Relative to our first session, the only
additional package that we need is <code>mixOmics</code>. This can be installed using
<code>BiocManager::install('mixOmics')</code>.</p>
<div id="interpreting-pls-da" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Interpreting PLS-DA<a class="anchor" aria-label="anchor" href="#interpreting-pls-da"><i class="fas fa-link"></i></a>
</h2>
<p>The T1D dataset below describes 427 metabolites from the gut microbiomes of 40
T1D patients and 61 healthy controls. These data were originally gathered by
<a href="https://doi.org/10.2337/dc18-0777">Gavin et al. (2018)</a>, who were motivated by
the relationship between the pancreas and intestinal issues often experienced by
T1D patients. They were especially curious about whether microbime-associated
proteins might be related to increased risk for T1D development.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span></span></code></pre></div>
<p>The output that we care about are the PLD-DA scores and loadings. The wrapper
below directly gives us this output without our having to explicitly set
hyerparameters, though you can <a href="https://github.com/krisrs1128/intro-to-simulation/blob/main/MIGsim/R/methods.R">look here</a>
to see how the function was defined.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MIGsim/man/splsda_fit.html">splsda_fit</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotIndiv.html">plotIndiv</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">fit</span>, cex <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/initial_t1d_analysis-1.png" width="672"></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotVar.html">plotVar</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">fit</span>, cutoff <span class="op">=</span> <span class="fl">0.4</span>, cex <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/initial_t1d_analysis-2.png" width="672"></div>
<p><strong>Exercise</strong>: Discuss the output of <code>plotIndiv</code>. How does <code>plotVar</code> shape
your interpretation?</p>
<p><strong>Solution</strong>: The first two dimensions of the sPLS-DA pick on distinctions between the T1D
(orange) and control (blue) groups. The variance by these first two dimensions
is relatively small
– variation in protein signatures may not be easiliy captured in two dimensions
or may not generally be correlated with the disease response. Since most of the
differences between groups are captured by the first dimension, we can look for
features that are highly correlated with this dimension to see which proteins
might be driving the difference. For example, we can conclude that T1D samples
tend to have higher levels of
<a href="https://www.uniprot.org/uniprotkb/P21333/entry">p21333</a> and
<a href="https://www.uniprot.org/uniprotkb/Q8WZ42/entry">q8wz42</a> compared to the
controls.</p>
</div>
<div id="estimation" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Estimation<a class="anchor" aria-label="anchor" href="#estimation"><i class="fas fa-link"></i></a>
</h2>
<p>Let’s estimate a simulator where every protein is allowed ot vary across T1D
type. Since the data have already been centered log-ratio transformed, it’s okay
to treat these as Gaussian.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulator</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/setup_simulator.html">setup_simulator</a></span><span class="op">(</span><span class="va">t1d</span>, link_formula <span class="op">=</span> <span class="op">~</span><span class="va">outcome2</span>, family <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/gamboostLSS/man/families.html">GaussianLSS</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="evaluation" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"><i class="fas fa-link"></i></a>
</h2>
<p>Last time, we saw how we could visualize marginal simulator quality. How can we
tell whether a joint simulator is working, though? One simple check is to
analyze the pairwise correlations. Since the copula model is designed to capture
second-order moments, it should at the very least effectively capture the
correlations.</p>
<p>We’ve written a small helper that visualizes the pairwise protein-protein
correlations from both the real and the simulated datasets. We seem to be often
overestimating the correlation strength. This is likely a consequence of the
high-dimensionality of the problem.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_exper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">simulator</span><span class="op">)</span></span>
<span><span class="fu">correlation_hist</span><span class="op">(</span><span class="va">t1d</span>, <span class="va">sim_exper</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/multivariate_correlations-1.png" width="672"></div>
<p><strong>Exercise</strong>: To address this, let’s try modifying the <code>copula_def</code> argument of
<code>setup_simulator</code> to use a more suitable simulator. Generate new correlation
histograms and comment on the changes you observe. You only need to modify the
commented lines (<code>#</code>) lines in the block below.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulator</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/setup_simulator.html">setup_simulator</a></span><span class="op">(</span></span>
<span>  <span class="va">t1d</span>,</span>
<span>  link_formula <span class="op">=</span> <span class="op">~</span><span class="va">outcome2</span>,</span>
<span>  family <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/gamboostLSS/man/families.html">GaussianLSS</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  copula_def <span class="op">=</span> <span class="co"># fill this code in</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim_exper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">simulator</span><span class="op">)</span> <span class="co"># and then run these two lines</span></span>
<span><span class="fu">correlation_hist</span><span class="op">(</span><span class="va">t1d</span>, <span class="va">sim_exper</span><span class="op">)</span> <span class="co">#</span></span></code></pre></div>
<p><strong>Solution</strong>: For our copula, we can use a covariance estimator of <a href="https://www.tandfonline.com/doi/abs/10.1198/jasa.2011.tm10560">Cai and Liu
(2011)</a>, that is
suited for high dimensions. Larger values of <code>thr</code> will increase the stability
of our estimates, but at the cost of potentially missing or weakening true
correlations. In line with this point, our new simulated correlations are more
concentrated.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulator</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/setup_simulator.html">setup_simulator</a></span><span class="op">(</span></span>
<span>  <span class="va">t1d</span>,</span>
<span>  link_formula <span class="op">=</span> <span class="op">~</span><span class="va">outcome2</span>,</span>
<span>  family <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/gamboostLSS/man/families.html">GaussianLSS</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  copula_def <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/copula_adaptive.html">copula_adaptive</a></span><span class="op">(</span>thr <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim_exper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">simulator</span><span class="op">)</span></span>
<span><span class="fu">correlation_hist</span><span class="op">(</span><span class="va">t1d</span>, <span class="va">sim_exper</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure">
<img src="simulation_for_microbiome_files/figure-html/estimate_t1d_simulator-1.png" width="672">
:::</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">contrast_boxplot</span><span class="op">(</span><span class="va">t1d</span>, <span class="va">sim_exper</span>, <span class="op">~</span><span class="va">.</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">outcome2</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/t1d_contrast_boxplot-1.png" width="672"></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rhos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>sim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">sim_exper</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, real <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">rhos</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">top_cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">rhos</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">rhos</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fl">0.999</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate data for the pairwise scatterplots</span></span>
<span><span class="va">pairs_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  real <span class="op">=</span> <span class="fu">subset_correlated</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span>, <span class="va">top_cors</span><span class="op">)</span>,</span>
<span>  simulated <span class="op">=</span> <span class="fu">subset_correlated</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">sim_exper</span><span class="op">)</span>, <span class="va">top_cors</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">keep_pairs</span> <span class="op">&lt;-</span> <span class="va">pairs_data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">source</span>, <span class="va">pair</span>, <span class="va">correlation</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">source</span> <span class="op">==</span> <span class="st">"real"</span>, <span class="va">correlation</span> <span class="op">&gt;</span> <span class="fl">0.72</span>, <span class="va">correlation</span> <span class="op">&lt;</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">pair</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize the pairwise scatterplots</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">pairs_data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pair</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">keep_pairs</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">feature1</span>, <span class="va">feature2</span>, col <span class="op">=</span> <span class="va">source</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#038C8C"</span>, <span class="st">"#D93636"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"A"</span>, col <span class="op">=</span> <span class="st">"Source"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="st">"Feature"</span> <span class="op">~</span> <span class="va">j</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="st">"Feature"</span> <span class="op">~</span> <span class="va">j</span><span class="op">^</span><span class="st">"'"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">pair</span>, <span class="op">-</span><span class="va">correlation</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">6</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,</span>
<span>    axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create the heatmap</span></span>
<span><span class="va">ft_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">order</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">correlation_heatmap</span><span class="op">(</span><span class="va">rhos</span><span class="op">$</span><span class="va">real</span>, <span class="va">ft_order</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"B  Real"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="op">(</span><span class="fu">correlation_heatmap</span><span class="op">(</span><span class="va">rhos</span><span class="op">$</span><span class="va">sim</span>, <span class="va">ft_order</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Simulated"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/collect_plots_multivariate-1.png" width="672"></div>
<p>How did we choose this threshold? You can compare the correlations more
systematically using this loop:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bivariate_metrics</span> <span class="op">&lt;-</span> \<span class="op">(</span><span class="va">thr</span>, <span class="va">n_rep</span> <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># refit the simulator with current threshold</span></span>
<span>  <span class="va">simulator</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/setup_simulator.html">setup_simulator</a></span><span class="op">(</span></span>
<span>    <span class="va">t1d</span>,</span>
<span>    link_formula <span class="op">=</span> <span class="op">~</span><span class="va">outcome2</span>,</span>
<span>    family <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/gamboostLSS/man/families.html">GaussianLSS</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    copula_def <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scDesigner/man/copula_adaptive.html">copula_adaptive</a></span><span class="op">(</span>thr <span class="op">=</span> <span class="va">thr</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># compute metrics on five samples</span></span>
<span>  <span class="va">err</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_rep</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sim_exper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">simulator</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span>, <span class="op">]</span></span>
<span>    <span class="va">rhos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>sim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">sim_exper</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, real <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">err</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>      frobenius <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="op">(</span><span class="va">rhos</span><span class="op">$</span><span class="va">sim</span> <span class="op">-</span> <span class="va">rhos</span><span class="op">$</span><span class="va">real</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      ks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">rhos</span><span class="op">$</span><span class="va">sim</span>, <span class="va">rhos</span><span class="op">$</span><span class="va">real</span><span class="op">)</span><span class="op">$</span><span class="va">statistic</span>,</span>
<span>      thr <span class="op">=</span> <span class="va">thr</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">err</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">errors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">.001</span>, <span class="fl">0.2</span>, length.out <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html">map_dfr</a></span><span class="op">(</span><span class="va">bivariate_metrics</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">thr</span>, names_to <span class="op">=</span> <span class="st">"metric"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">errors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">thr</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">metric</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Threshold"</span>, y <span class="op">=</span> <span class="st">"Metric Value"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/metrics_results-1.png" width="672"></div>
</div>
<div id="pls-da-power-analysis" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> PLS-DA Power Analysis<a class="anchor" aria-label="anchor" href="#pls-da-power-analysis"><i class="fas fa-link"></i></a>
</h2>
<p>Now that we have a simulator, we can run a power analysis. In theory, we could
look at how any summary from the PLS-DA output varies as the sample size
increases. The most natural one, though, is simply to see how classifier
performance improves as we gather more samples. Specifically, we’ll measure the
holdout Area Under the Curve (auc), a measure of how well the trains PLS-DA
classifier balance precision and recall on new samples.</p>
<p>Moreover, we’ll study the effect of sparsity – what happens when many features
have no relationship at all with the response? We’ll also simulate three
hypothetical datasets for each sample size and sparsity level. All
configurations of interest are stored in the <code>config</code> matrix below.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span></span>
<span>  sample_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">200</span>, length.out <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  n_rep <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>  n_null <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">317</span>, <span class="fl">417</span>, length.out <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  metrics <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">t1d_order</span><span class="op">)</span></span></code></pre></div>
<p><strong>Exercise</strong>: Finally, we’re in a position to generate synthetic data and
evaluate PLS-DA performance. Fill in the block below to update the simulator for
each <code>i</code>. Remember that the original <code>simulator</code> defined above assumes that all
proteins are associated with T1D. You can use <code>t1d_order</code> to prioritize the
proteins with the strongest effects in the original data. As before, you should
only need to modify the line marked with the comments (<code>#</code>).</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">config</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">simulator</span> <span class="op">&lt;-</span> <span class="va">simulator</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>      <span class="co"># fill this in</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">config</span><span class="op">$</span><span class="va">metrics</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="va">simulator</span>, <span class="va">config</span><span class="op">$</span><span class="va">sample_size</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/MIGsim/man/splsda_fit.html">splsda_fit</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="st">"auc"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"run {i}/{nrow(config)}"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><strong>Solution</strong>: To define nulls, we mutate the weakest proteins so that there is no longer any
association with T1D: <code>link = ~ 1</code> instead of <code>link = ~ outcome2</code>. To speed up
the computation, we organized the <code>mutate</code> calls so that we don’t need to
re-estimate proteins whose effects were removed in a previous iteration.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">config</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">||</span> <span class="va">config</span><span class="op">$</span><span class="va">n_null</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">!=</span> <span class="va">config</span><span class="op">$</span><span class="va">n_null</span><span class="op">[</span><span class="va">i</span> <span class="op">-</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">simulator</span> <span class="op">&lt;-</span> <span class="va">simulator</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">scDesigner</span><span class="fu">::</span><span class="fu">mutate</span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">any_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="va">t1d_order</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">config</span><span class="op">$</span><span class="va">n_null</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, link <span class="op">=</span> <span class="op">~</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">estimate</span><span class="op">(</span>mstop <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="va">config</span><span class="op">$</span><span class="va">metrics</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="va">simulator</span>, <span class="va">config</span><span class="op">$</span><span class="va">sample_size</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/MIGsim/man/splsda_fit.html">splsda_fit</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="st">"auc"</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html">glue</a></span><span class="op">(</span><span class="st">"run {i}/{nrow(config)}"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://uwmadison.box.com/shared/static/tf1xyo7a2n2rd2ms4y9itaev42dsnp5c.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>:::</p>
<p>We can visualize variation in performance.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">config</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sample_size</span><span class="op">)</span>, <span class="va">metrics</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">n_null</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="http://mjskay.github.io/ggdist/reference/stat_pointinterval.html">stat_pointinterval</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scico/man/ggplot2-scales.html">scale_color_scico_d</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"managua"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sample Size"</span>, y <span class="op">=</span> <span class="st">"AUC"</span>, color <span class="op">=</span> <span class="st">"# Nulls"</span>, title <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="simulation_for_microbiome_files/figure-html/visualize-power-1.png" width="672"></div>
<p><strong>Discussion</strong>: Interpret the visualization above. How do you think analysis
like this could help you justify making some experimental investments over
others?</p>
<p><strong>Solution</strong>: Reading across facets from top left to bottom right, power decreases when the
number of null proteins increases. It seems that sPLS-DA can benefit from having
many weakly associated features. While power is sometimes high in low sample
sizes, the variance can be quite large. In all settings, there is a noticeable
decrease in variance in power as we go from 15 to 48 samples. If we can assume
that a moderate fraction (&gt; 15%) of measured proteins are associated with T1D,
then we may already achieve good power with ~ 100 samples. However, if we
imagine our effect might be sparser in our future experiment, then this figure
would give us good justification for arguing for a larger number of samples, in
order to ensure we can discover a disease-associated proteomics signature.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">$</span><span class="va">outcome2</span> <span class="op">==</span> <span class="st">"T1D"</span></span>
<span><span class="va">p_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span>, <span class="fl">1</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="op">-</span><span class="va">ix</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/mutoss/man/storey_pi0_est.html">storey_pi0_est</a></span><span class="op">(</span><span class="va">p_vals</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">$</span><span class="va">pi0</span></span></code></pre></div>
<pre><code>## [1] 1.311475</code></pre>
</div>
<div id="comparison-with-t-test-calculation" class="section level2" number="3.5">
<h2>
<span class="header-section-number">3.5</span> Comparison with <span class="math inline">\(t\)</span>-test calculation<a class="anchor" aria-label="anchor" href="#comparison-with-t-test-calculation"><i class="fas fa-link"></i></a>
</h2>
<p>What if we had based our power analysis using theory from <span class="math inline">\(t\)</span>-tests? We might be
hopeful that this theory might give a sample size that is also good enough for
the sPLS-DA. To make this concrete, suppose that we will test all 427 taxa using
a two-sample <span class="math inline">\(t\)</span>-test with a Benjamini-Hochberg correction for multiple testing.
Suppose that we are in the sparse case, with only 10 nonnull taxa. As a
hypothetical true signal strength, we use the observed signal strength of the
tenth most significant taxon from the real data,</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset_size</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">$</span><span class="va">outcome2</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">[</span><span class="va">t1d_order</span><span class="op">[</span><span class="va">subset_size</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="va">test_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">y</span> <span class="op">==</span> <span class="st">"T1D"</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="va">y</span> <span class="op">==</span> <span class="st">"healthy"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">signal_strength</span> <span class="op">&lt;-</span> <span class="va">test_result</span><span class="op">$</span><span class="va">statistic</span></span>
<span><span class="va">signal_strength</span></span></code></pre></div>
<pre><code>##         t 
## -2.092196</code></pre>
<p>How many samples would we need in order to achieve 80% power in this setting?
Accounting for multiple testing, we would need this taxon to be significant at
the level <span class="math inline">\(\frac{10 \alpha}{427}\)</span>, which gives us,</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pwr/man/pwr.t.test.html">pwr.t.test</a></span><span class="op">(</span>power <span class="op">=</span> <span class="fl">0.8</span>, d <span class="op">=</span> <span class="va">signal_strength</span>, sig.level <span class="op">=</span> <span class="fl">0.05</span> <span class="op">*</span> <span class="va">subset_size</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">t1d</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">n</span></span></code></pre></div>
<pre><code>## [1] 10.36249</code></pre>
<p>So, according to this power analysis, we only need 11 samples each for the
healthy and disease groups. In comparison to our earlier analysis using sPLS-DA,
this is very optimistic. In the case where only 10 taxa are significant, then
when we had 20 samples total, the AUC was in the range <span class="math inline">\(\left[\right]\)</span>, which is
only slightly better than random guessing. This is a case where a <span class="math inline">\(t\)</span>-test power
calculation would make us overoptimistic about the number of samples that are
required. Both the data generation (two shifted normal distributions) and
analysis (<span class="math inline">\(t\)</span>-tests with BH correction) mechanisms assumed by this power
analysis are suspect, and in this case it makes a difference. All power analysis
requires some simplification, but the fact that a more realistic simulator
points us towards noticeably larger sample sizes is an indication that we should
not trust the <span class="math inline">\(t\)</span>-test calculation and instead suggest the larger sample sample
from our sPLS-DA specific analysis.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>## R version 4.4.1 Patched (2024-08-21 r87049)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sonoma 14.5
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Chicago
## tzcode source: internal
## 
## attached base packages:
##  [1] splines   parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] doRNG_1.8.6                     rngtools_1.5.2                  foreach_1.5.2                   lubridate_1.9.3                 readr_2.1.5                    
##  [6] tidyverse_2.0.0                 ggpubr_0.6.0                    ggrepel_0.9.6                   edgeR_4.3.16                    limma_3.61.10                  
## [11] ANCOMBC_2.7.0                   stringr_1.5.1                   ks_1.14.3                       tidySummarizedExperiment_1.15.1 ttservice_0.4.1                
## [16] SimBench_0.99.1                 wesanderson_0.3.7               mia_1.13.36                     MultiAssayExperiment_1.31.5     TreeSummarizedExperiment_2.13.0
## [21] Biostrings_2.73.1               XVector_0.45.0                  SingleCellExperiment_1.27.2     gamlss_5.4-22                   nlme_3.1-166                   
## [26] gamlss.dist_6.1-1               gamlss.data_6.0-6               scDesigner_0.0.0.9000           purrr_1.0.2                     MIGsim_0.0.0.9000              
## [31] tidyr_1.3.1                     tibble_3.2.1                    scico_1.5.0                     pwr_1.3-0                       patchwork_1.3.0                
## [36] mutoss_0.1-13                   mvtnorm_1.3-1                   mixOmics_6.29.0                 lattice_0.22-6                  MASS_7.3-61                    
## [41] glue_1.8.0                      ggplot2_3.5.1                   ggdist_3.3.2                    gamboostLSS_2.0-7               mboost_2.9-11                  
## [46] stabs_0.6-4                     forcats_1.0.0                   dplyr_1.1.4                     SummarizedExperiment_1.35.1     Biobase_2.65.1                 
## [51] GenomicRanges_1.57.1            GenomeInfoDb_1.41.1             IRanges_2.39.2                  S4Vectors_0.43.2                BiocGenerics_0.51.1            
## [56] MatrixGenerics_1.17.0           matrixStats_1.4.1               SpiecEasi_1.1.3                 CovTools_0.5.4                 
## 
## loaded via a namespace (and not attached):
##   [1] igraph_2.0.3                ica_1.0-3                   plotly_4.10.4               Formula_1.2-5               scater_1.33.4               zlibbioc_1.51.1            
##   [7] tidyselect_1.2.1            bit_4.5.0                   doParallel_1.0.17           pspline_1.0-20              blob_1.2.4                  S4Arrays_1.5.8             
##  [13] png_0.1-8                   plotrix_3.8-4               cli_3.6.3                   CVXR_1.0-14                 pulsar_0.3.11               multtest_2.61.0            
##  [19] goftest_1.2-3               textshaping_0.4.0           bluster_1.15.1              BiocNeighbors_1.99.1        uwot_0.2.2                  mime_0.12                  
##  [25] evaluate_1.0.0              tidytree_0.4.6              leiden_0.4.3.1              stringi_1.8.4               backports_1.5.0             lmerTest_3.1-3             
##  [31] Exact_3.3                   gsl_2.1-8                   httpuv_1.6.15               AnnotationDbi_1.67.0        magrittr_2.0.3              mclust_6.1.1               
##  [37] ADGofTest_0.3               pcaPP_2.0-5                 rootSolve_1.8.2.4           sctransform_0.4.1           lpSolve_5.6.21              ggbeeswarm_0.7.2           
##  [43] DBI_1.2.3                   jquerylib_0.1.4             withr_3.0.1                 corpcor_1.6.10              systemfonts_1.1.0           class_7.3-22               
##  [49] lmtest_0.9-40               htmlwidgets_1.6.4           fs_1.6.4                    labeling_0.4.3              SparseArray_1.5.38          cellranger_1.1.0           
##  [55] DESeq2_1.45.3               extrafont_0.19              lmom_3.0                    flare_1.7.0.1               reticulate_1.39.0           minpack.lm_1.2-4           
##  [61] geigen_2.3                  zoo_1.8-12                  knitr_1.48                  nnls_1.5                    UCSC.utils_1.1.0            SHT_0.1.8                  
##  [67] timechange_0.3.0            decontam_1.25.0             fansi_1.0.6                 grid_4.4.1                  data.table_1.16.0           vegan_2.6-8                
##  [73] RSpectra_0.16-2             irlba_2.3.5.1               extrafontdb_1.0             DescTools_0.99.56           ellipsis_0.3.2              fastDummies_1.7.4          
##  [79] lazyeval_0.2.2              yaml_2.3.10                 survival_3.7-0              scattermore_1.2             crayon_1.5.3                mediation_4.5.0            
##  [85] RcppAnnoy_0.0.22            RColorBrewer_1.1-3          progressr_0.14.0            later_1.3.2                 ggridges_0.5.6              codetools_0.2-20           
##  [91] base64enc_0.1-3             Seurat_5.1.0                KEGGREST_1.45.1             Rtsne_0.17                  shape_1.4.6.1               randtoolbox_2.0.4          
##  [97] foreign_0.8-87              pkgconfig_2.0.3             xml2_1.3.6                  spatstat.univar_3.0-1       downlit_0.4.4               scatterplot3d_0.3-44       
## [103] spatstat.sparse_3.1-0       ape_5.8                     viridisLite_0.4.2           xtable_1.8-4                highr_0.11                  car_3.1-2                  
## [109] fastcluster_1.2.6           plyr_1.8.9                  httr_1.4.7                  rbibutils_2.2.16            tools_4.4.1                 globals_0.16.3             
## [115] SeuratObject_5.0.2          kde1d_1.0.7                 beeswarm_0.4.0              htmlTable_2.4.3             broom_1.0.6                 checkmate_2.3.2            
## [121] rgl_1.3.1                   assertthat_0.2.1            lme4_1.1-35.5               rngWELL_0.10-9              digest_0.6.37               permute_0.9-7              
## [127] numDeriv_2016.8-1.1         bookdown_0.40               Matrix_1.7-0                tzdb_0.4.0                  farver_2.1.2                reshape2_1.4.4             
## [133] yulab.utils_0.1.7           viridis_0.6.5               DirichletMultinomial_1.47.0 rpart_4.1.23                cachem_1.1.0                polyclip_1.10-7            
## [139] WGCNA_1.73                  Hmisc_5.1-3                 generics_0.1.3              dynamicTreeCut_1.63-1       parallelly_1.38.0           statmod_1.5.0              
## [145] impute_1.79.0               huge_1.3.5                  ragg_1.3.3                  RcppHNSW_0.6.0              ScaledMatrix_1.13.0         carData_3.0-5              
## [151] minqa_1.2.8                 pbapply_1.7-2               glmnet_4.1-8                spam_2.10-0                 utf8_1.2.4                  gtools_3.9.5               
## [157] shapes_1.2.7                readxl_1.4.3                preprocessCore_1.67.0       inum_1.0-5                  ggsignif_0.6.4              energy_1.7-12              
## [163] gridExtra_2.3               shiny_1.9.1                 GenomeInfoDbData_1.2.12     memoise_2.0.1               rmarkdown_2.28              scales_1.3.0               
## [169] gld_2.6.6                   stabledist_0.7-2            partykit_1.2-22             svglite_2.1.3               future_1.34.0               RANN_2.6.2                 
## [175] spatstat.data_3.1-2         rstudioapi_0.16.0           cluster_2.1.6               spatstat.utils_3.1-0        hms_1.1.3                   fitdistrplus_1.2-1         
## [181] munsell_0.5.1               cowplot_1.1.3               colorspace_2.1-1            ellipse_0.5.0               rlang_1.1.4                 quadprog_1.5-8             
## [187] copula_1.1-4                DelayedMatrixStats_1.27.3   sparseMatrixStats_1.17.2    dotCall64_1.1-1             scuttle_1.15.4              mgcv_1.9-1                 
## [193] xfun_0.47                   e1071_1.7-16                TH.data_1.1-2               iterators_1.0.14            rARPACK_0.11-0              abind_1.4-8                
## [199] libcoin_1.0-10              treeio_1.29.1               gmp_0.7-5                   DECIPHER_3.1.4              Rdpack_2.6.1                promises_1.3.0             
## [205] RSQLite_2.3.7               sandwich_3.1-1              proxy_0.4-27                DelayedArray_0.31.11        Rmpfr_0.9-5                 GO.db_3.20.0               
## [211] compiler_4.4.1              prettyunits_1.2.0           boot_1.3-31                 distributional_0.5.0        beachmat_2.21.6             rbiom_1.0.3                
## [217] listenv_0.9.1               Rcpp_1.0.13                 Rttf2pt1_1.3.12             BiocSingular_1.21.4         tensor_1.5                  progress_1.2.3             
## [223] BiocParallel_1.39.0         insight_0.20.5              spatstat.random_3.3-2       R6_2.5.1                    fastmap_1.2.0               multcomp_1.4-26            
## [229] rstatix_0.7.2               vipor_0.4.7                 ROCR_1.0-11                 rsvd_1.0.5                  nnet_7.3-19                 gtable_0.3.5               
## [235] KernSmooth_2.23-24          miniUI_0.1.1.1              deldir_2.0-4                htmltools_0.5.8.1           ggthemes_5.1.0              RcppParallel_5.1.9         
## [241] bit64_4.5.2                 spatstat.explore_3.3-2      lifecycle_1.0.4             nloptr_2.1.1                sass_0.4.9                  vctrs_0.6.5                
## [247] VGAM_1.1-12                 slam_0.1-53                 spatstat.geom_3.3-3         sp_2.1-4                    future.apply_1.11.2         pracma_2.4.4               
## [253] rvinecopulib_0.6.3.1.1      bslib_0.8.0                 pillar_1.9.0                locfit_1.5-9.10             jsonlite_1.8.9              expm_1.0-0</code></pre>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="simulating-differential-abundance.html"><span class="header-section-number">2</span> Simulating Differential Abundance</a></div>
<div class="next"><a href="microbiome-networks.html"><span class="header-section-number">4</span> Microbiome Networks</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#multivariate-power-analysis"><span class="header-section-number">3</span> Multivariate Power Analysis</a></li>
<li><a class="nav-link" href="#interpreting-pls-da"><span class="header-section-number">3.1</span> Interpreting PLS-DA</a></li>
<li><a class="nav-link" href="#estimation"><span class="header-section-number">3.2</span> Estimation</a></li>
<li><a class="nav-link" href="#evaluation"><span class="header-section-number">3.3</span> Evaluation</a></li>
<li><a class="nav-link" href="#pls-da-power-analysis"><span class="header-section-number">3.4</span> PLS-DA Power Analysis</a></li>
<li><a class="nav-link" href="#comparison-with-t-test-calculation"><span class="header-section-number">3.5</span> Comparison with \(t\)-test calculation</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Simulation for Microbiome Analysis</strong>" was written by Kris Sankaran, Saritha Kodikara, Jingyi Jessica Li, and Kim-Anh Lê Cao. It was last built on 2024-10-08.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
