<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Vertical Integration | Simulation for Microbiome Analysis</title>
  <meta name="description" content="Examples illustrating the use of simulation for microbiome analysis, companion to the review article:" />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Vertical Integration | Simulation for Microbiome Analysis" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Examples illustrating the use of simulation for microbiome analysis, companion to the review article:" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Vertical Integration | Simulation for Microbiome Analysis" />
  
  <meta name="twitter:description" content="Examples illustrating the use of simulation for microbiome analysis, companion to the review article:" />
  

<meta name="author" content="Kris Sankaran, Saritha Kodikara, and Kim-Anh LÃª Cao" />


<meta name="date" content="2024-07-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="batch-effect-correction.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Simulation for Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#setup"><i class="fa fa-check"></i><b>1.1</b> Setup</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#using-summarizedexperiment"><i class="fa fa-check"></i><b>1.2</b> Using <code>SummarizedExperiment</code></a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#warm-up-a-gaussian-example"><i class="fa fa-check"></i><b>1.3</b> Warm-up: A Gaussian Example</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="simulating-differential-abundance.html"><a href="simulating-differential-abundance.html"><i class="fa fa-check"></i><b>2</b> Simulating Differential Abundance</a>
<ul>
<li class="chapter" data-level="2.1" data-path="simulating-differential-abundance.html"><a href="simulating-differential-abundance.html#critique"><i class="fa fa-check"></i><b>2.1</b> Critique</a></li>
<li class="chapter" data-level="2.2" data-path="simulating-differential-abundance.html"><a href="simulating-differential-abundance.html#power-analysis-loop"><i class="fa fa-check"></i><b>2.2</b> Power Analysis Loop</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html"><i class="fa fa-check"></i><b>3</b> Multivariate Power Analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html#interpreting-pls-da"><i class="fa fa-check"></i><b>3.1</b> Interpreting PLS-DA</a></li>
<li class="chapter" data-level="3.2" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html#estimation"><i class="fa fa-check"></i><b>3.2</b> Estimation</a></li>
<li class="chapter" data-level="3.3" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html#evaluation"><i class="fa fa-check"></i><b>3.3</b> Evaluation</a></li>
<li class="chapter" data-level="3.4" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html#pls-da-power-analysis"><i class="fa fa-check"></i><b>3.4</b> PLS-DA Power Analysis</a></li>
<li class="chapter" data-level="3.5" data-path="multivariate-power-analysis.html"><a href="multivariate-power-analysis.html#comparison-with-t-test-calculation"><i class="fa fa-check"></i><b>3.5</b> Comparison with <span class="math inline">\(t\)</span>-test calculation</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="microbiome-networks.html"><a href="microbiome-networks.html"><i class="fa fa-check"></i><b>4</b> Microbiome Networks</a>
<ul>
<li class="chapter" data-level="4.1" data-path="microbiome-networks.html"><a href="microbiome-networks.html#estimation-1"><i class="fa fa-check"></i><b>4.1</b> Estimation</a></li>
<li class="chapter" data-level="4.2" data-path="microbiome-networks.html"><a href="microbiome-networks.html#evaluation-1"><i class="fa fa-check"></i><b>4.2</b> Evaluation</a></li>
<li class="chapter" data-level="4.3" data-path="microbiome-networks.html"><a href="microbiome-networks.html#block-covariance"><i class="fa fa-check"></i><b>4.3</b> Block Covariance</a></li>
<li class="chapter" data-level="4.4" data-path="microbiome-networks.html"><a href="microbiome-networks.html#generalization"><i class="fa fa-check"></i><b>4.4</b> Generalization</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="batch-effect-correction.html"><a href="batch-effect-correction.html"><i class="fa fa-check"></i><b>5</b> Batch Effect Correction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="batch-effect-correction.html"><a href="batch-effect-correction.html#anaerobic-digestion-data"><i class="fa fa-check"></i><b>5.1</b> Anaerobic Digestion Data</a></li>
<li class="chapter" data-level="5.2" data-path="batch-effect-correction.html"><a href="batch-effect-correction.html#simulation-design"><i class="fa fa-check"></i><b>5.2</b> Simulation Design</a></li>
<li class="chapter" data-level="5.3" data-path="batch-effect-correction.html"><a href="batch-effect-correction.html#benchmarking"><i class="fa fa-check"></i><b>5.3</b> Benchmarking</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="vertical-integration.html"><a href="vertical-integration.html"><i class="fa fa-check"></i><b>6</b> Vertical Integration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="vertical-integration.html"><a href="vertical-integration.html#icu-dataset"><i class="fa fa-check"></i><b>6.1</b> ICU Dataset</a></li>
<li class="chapter" data-level="6.2" data-path="vertical-integration.html"><a href="vertical-integration.html#interlude-using-map"><i class="fa fa-check"></i><b>6.2</b> Interlude: Using map</a></li>
<li class="chapter" data-level="6.3" data-path="vertical-integration.html"><a href="vertical-integration.html#simulation-question"><i class="fa fa-check"></i><b>6.3</b> Simulation Question</a></li>
<li class="chapter" data-level="6.4" data-path="vertical-integration.html"><a href="vertical-integration.html#simulation-results"><i class="fa fa-check"></i><b>6.4</b> Simulation Results</a></li>
<li class="chapter" data-level="6.5" data-path="vertical-integration.html"><a href="vertical-integration.html#alignability"><i class="fa fa-check"></i><b>6.5</b> Alignability</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Simulation for Microbiome Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="vertical-integration" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Vertical Integration<a href="vertical-integration.html#vertical-integration" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In horizontal integration, we have many datasets, all with the same features.
They only differ because they were gathered at different times. In contrast, for
vertical integration, we instead have many datasets all with the same <em>samples</em>.
They differ because they measure different aspects of those samples. Our goal in
this situation is not to remove differences across datasets, like it was in
horizontal integration, but instead to clarify the relationships across sources.</p>
<p>One important question that often arises in vertical integration is â are the
data even alignable? That is, in our effort to look for relationships across
datasets, we might accidentally miss out on interesting variation that exists
within the individual assays. If the technologies are measuring very different
things, we might be better off simply analyzing the data separately. To help us
gauge which setting we might be in, we can simulate data where we know that we
shouldnât align the sources. If our integration methods are giving similar
outputs as they give on this simulated data, then we should be more cautious.</p>
<p>There are a few ways in which a dataset might not be ``alignable.ââ The most
general reason is that there may be no latent sources of variation in common
between the sources. A simpler reason is that something that influenced one
assay substantially (e.g., disease state) might not influence the other by much.
Letâs see how an integration method might work in this setting.</p>
<div id="icu-dataset" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> ICU Dataset<a href="vertical-integration.html#icu-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Weâll work with the ICU sepsis dataset previously studied by <a href="https://doi.org/10.1128/mSystems.01148-20">Haak et al.
(2021)</a> and documented within a
<a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/microbiome_vignette.html#train-mofa-model">vignette</a>
for the MOFA package. The three datasets here are 16S bacterial, ITS fungal, and
Virome assays, all applied to different healthy and sepsis patient populations.
Moreover, some participants were on a course of antibiotics while others were
not. The question is how either sepsis, antibiotics, or their interaction
affects the microbiome viewed through these three assays. The data are printed
below, they have already been filtered and CLR transformed following the MOFA
vignette.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="vertical-integration.html#cb92-1" tabindex="-1"></a><span class="fu">data</span>(icu)</span>
<span id="cb92-2"><a href="vertical-integration.html#cb92-2" tabindex="-1"></a>icu</span></code></pre></div>
<pre><code>## $Bacteria
## # A SummarizedExperiment-tibble abstraction: 10,260 Ã 19
## # [90mFeatures=180 | Samples=57 | Assays=clr[0m
##    .feature        .sample    clr   Age Sexs   Diagnosis Category
##    &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt; &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;   
##  1 Acidaminococcus TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  2 Actinomyces     TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  3 Adlercreutzia   TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  4 Agathobacter    TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  5 Akkermansia     TKI_F1   1.68     89 Female Sepsis, â¦ Sepsis  
##  6 Alistipes       TKI_F1  10.9      89 Female Sepsis, â¦ Sepsis  
##  7 Allisonella     TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  8 Alloprevotella  TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
##  9 Anaerofilum     TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
## 10 Anaerofustis    TKI_F1  -0.717    89 Female Sepsis, â¦ Sepsis  
## # â¹ 40 more rows
## # â¹ 12 more variables: Penicillins &lt;lgl&gt;, Cephalosporins &lt;lgl&gt;,
## #   Carbapenems &lt;lgl&gt;, Macrolides &lt;lgl&gt;, Aminoglycosides &lt;lgl&gt;,
## #   Quinolones &lt;lgl&gt;, Co_trimoxazole &lt;lgl&gt;, Metronidazole &lt;lgl&gt;,
## #   Vancomycin &lt;lgl&gt;, Acetate &lt;dbl&gt;, Propionate &lt;dbl&gt;,
## #   Butyrate &lt;dbl&gt;
## 
## $Fungi
## # A SummarizedExperiment-tibble abstraction: 1,026 Ã 19
## # [90mFeatures=18 | Samples=57 | Assays=clr[0m
##    .feature      .sample    clr   Age Sexs   Diagnosis   Category
##    &lt;chr&gt;         &lt;chr&gt;    &lt;dbl&gt; &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt;   
##  1 Agaricus      TKI_F1  -1.21     89 Female Sepsis, puâ¦ Sepsis  
##  2 Aspergillus   TKI_F1   6.63     89 Female Sepsis, puâ¦ Sepsis  
##  3 Aureobasidium TKI_F1  -4.25     89 Female Sepsis, puâ¦ Sepsis  
##  4 Candida       TKI_F1   8.16     89 Female Sepsis, puâ¦ Sepsis  
##  5 Cladosporium  TKI_F1   1.05     89 Female Sepsis, puâ¦ Sepsis  
##  6 Debaryomyces  TKI_F1  -0.540    89 Female Sepsis, puâ¦ Sepsis  
##  7 Dipodascus    TKI_F1  -1.86     89 Female Sepsis, puâ¦ Sepsis  
##  8 Filobasidium  TKI_F1  -1.21     89 Female Sepsis, puâ¦ Sepsis  
##  9 Issatchenkia  TKI_F1   0.141    89 Female Sepsis, puâ¦ Sepsis  
## 10 Malassezia    TKI_F1   0.764    89 Female Sepsis, puâ¦ Sepsis  
## # â¹ 350 more rows
## # â¹ 12 more variables: Penicillins &lt;lgl&gt;, Cephalosporins &lt;lgl&gt;,
## #   Carbapenems &lt;lgl&gt;, Macrolides &lt;lgl&gt;, Aminoglycosides &lt;lgl&gt;,
## #   Quinolones &lt;lgl&gt;, Co_trimoxazole &lt;lgl&gt;, Metronidazole &lt;lgl&gt;,
## #   Vancomycin &lt;lgl&gt;, Acetate &lt;dbl&gt;, Propionate &lt;dbl&gt;,
## #   Butyrate &lt;dbl&gt;
## 
## $Viruses
## # A SummarizedExperiment-tibble abstraction: 2,394 Ã 19
## # [90mFeatures=42 | Samples=57 | Assays=clr[0m
##    .feature         .sample    clr   Age Sexs  Diagnosis Category
##    &lt;chr&gt;            &lt;chr&gt;    &lt;dbl&gt; &lt;int&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;   
##  1 Acidovorax phage TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  2 Acinetobacter pâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  3 Aeromonas phage  TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  4 Anellovirus      TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  5 Arthrobacter phâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  6 Bacillus phage   TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  7 Bacteroides phaâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  8 Bacteroidetes pâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
##  9 Brochothrix phaâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
## 10 Burkholderia phâ¦ TKI_F1  -0.602    89 Femaâ¦ Sepsis, â¦ Sepsis  
## # â¹ 32 more rows
## # â¹ 12 more variables: Penicillins &lt;lgl&gt;, Cephalosporins &lt;lgl&gt;,
## #   Carbapenems &lt;lgl&gt;, Macrolides &lt;lgl&gt;, Aminoglycosides &lt;lgl&gt;,
## #   Quinolones &lt;lgl&gt;, Co_trimoxazole &lt;lgl&gt;, Metronidazole &lt;lgl&gt;,
## #   Vancomycin &lt;lgl&gt;, Acetate &lt;dbl&gt;, Propionate &lt;dbl&gt;,
## #   Butyrate &lt;dbl&gt;</code></pre>
<p>We can simultaneously analyze these data sources using block sPLS-DA. This is
the multi-assay version of the analysis that we saw in the previous session.
<code>exper_splsda</code> is a very light wrapper of a mixOmics function call, which you
can read
<a href="https://github.com/krisrs1128/intro-to-simulation/blob/9f842ef52fdbe9a137833531147ceb3bc1f7ae81/MIGsim/R/vertical_integration.R#L3">here</a>.
The output plot below shows that each assay differs across groups, and this is
quantitatively summarized by the high estimated weights between each category
and the estimated PLS directions.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="vertical-integration.html#cb94-1" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">exper_splsda</span>(icu)</span>
<span id="cb94-2"><a href="vertical-integration.html#cb94-2" tabindex="-1"></a><span class="fu">plotIndiv</span>(fit)</span></code></pre></div>
<p><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-33-1.png" width="672" /></p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="vertical-integration.html#cb95-1" tabindex="-1"></a>fit<span class="sc">$</span>weights</span></code></pre></div>
<pre><code>##              comp1     comp2
## Bacteria 0.8580402 0.8191458
## Fungi    0.6513668 0.4845247
## Viruses  0.6118091 0.8047067</code></pre>
</div>
<div id="interlude-using-map" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Interlude: Using map<a href="vertical-integration.html#interlude-using-map" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the examples below, weâll find it helpful to use the function <code>map</code> in the
purrr package. This function gives a one-line replacement for simple for-loops;
it is analogous to list comprehensions in python. It can be useful many places
besides the topic of this tutorial. For example, if we want to convert the
vector <code>c(1, 2, 3)</code> into <code>c(1, 4, 9)</code>, we can use this map:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="vertical-integration.html#cb97-1" tabindex="-1"></a><span class="fu">map</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="sc">~</span> .<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [[1]]
## [1] 1
## 
## [[2]]
## [1] 4
## 
## [[3]]
## [1] 9</code></pre>
<p>The <code>~</code> notation is shorthand for defining a function, and the <code>.</code> represents
the current vector element. More generally, we can apply map to lists. This
line will update the list so that 1 is added to each element.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="vertical-integration.html#cb99-1" tabindex="-1"></a><span class="fu">map</span>(<span class="fu">list</span>(<span class="at">a =</span> <span class="dv">1</span>, <span class="at">b =</span> <span class="dv">2</span>, <span class="at">c =</span> <span class="dv">3</span>), <span class="sc">~</span> . <span class="sc">+</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## $a
## [1] 2
## 
## $b
## [1] 3
## 
## $c
## [1] 4</code></pre>
<p><strong>Exercise</strong>: To test your understanding, can you write a map that computes the
*mean for each
vector in the list <code>x</code> below? What about the mean of the 10 smallest elements?</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="vertical-integration.html#cb101-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">a =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>), <span class="at">b =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">1</span>))</span></code></pre></div>
<button onclick="myFunction(&#39;q-map&#39;)">
Show solution
</button>
<div id="q-map" style="display:none">
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="vertical-integration.html#cb102-1" tabindex="-1"></a><span class="fu">map</span>(x, mean)</span></code></pre></div>
<pre><code>## $a
## [1] 0.1177328
## 
## $b
## [1] 1.128995</code></pre>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="vertical-integration.html#cb104-1" tabindex="-1"></a><span class="fu">map</span>(x, <span class="sc">~</span> <span class="fu">mean</span>(<span class="fu">sort</span>(.)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]))</span></code></pre></div>
<pre><code>## $a
## [1] -1.920247
## 
## $b
## [1] -0.8639795</code></pre>
</div>
</div>
<div id="simulation-question" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Simulation Question<a href="vertical-integration.html#simulation-question" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>How would the output have looked if 16S community composition had not been
related to disease or antibiotics groups? Since integrative analysis prioritizes
similarities across sources, we expect this to mask some of the real differences
in the fungal and virus data as well. We can use simulation to gauge the extent
of this masking.</p>
<p>Our first step is to train a simulator. Weâre just learning four different
setes of parameters for each of the four observed groups. This is not as nuanced
as learning separate effects for sepsis and antibiotics, but it will be enough
for illustration. We have used <code>map</code> to estimate a simulator for each assay in
the <code>icu</code> list.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="vertical-integration.html#cb106-1" tabindex="-1"></a>simulator <span class="ot">&lt;-</span> <span class="fu">map</span>(</span>
<span id="cb106-2"><a href="vertical-integration.html#cb106-2" tabindex="-1"></a>  icu,</span>
<span id="cb106-3"><a href="vertical-integration.html#cb106-3" tabindex="-1"></a>  <span class="sc">~</span> <span class="fu">setup_simulator</span>(., <span class="sc">~</span>Category, <span class="sc">~</span> <span class="fu">GaussianLSS</span>()) <span class="sc">|&gt;</span></span>
<span id="cb106-4"><a href="vertical-integration.html#cb106-4" tabindex="-1"></a>    <span class="fu">estimate</span>(<span class="at">nu =</span> <span class="fl">0.05</span>)</span>
<span id="cb106-5"><a href="vertical-integration.html#cb106-5" tabindex="-1"></a>)</span></code></pre></div>
<p>So far, we havenât tried removing any relationships present in the 16S assay,
and indeed our integrative analysis output on the simulated data looks
comparable to that from the original study.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="vertical-integration.html#cb107-1" tabindex="-1"></a>icu_sim <span class="ot">&lt;-</span> <span class="fu">join_copula</span>(simulator, <span class="fu">copula_adaptive</span>()) <span class="sc">|&gt;</span></span>
<span id="cb107-2"><a href="vertical-integration.html#cb107-2" tabindex="-1"></a>  <span class="fu">sample</span>() <span class="sc">|&gt;</span></span>
<span id="cb107-3"><a href="vertical-integration.html#cb107-3" tabindex="-1"></a>  <span class="fu">split_assays</span>()</span>
<span id="cb107-4"><a href="vertical-integration.html#cb107-4" tabindex="-1"></a></span>
<span id="cb107-5"><a href="vertical-integration.html#cb107-5" tabindex="-1"></a>fit_sim <span class="ot">&lt;-</span> <span class="fu">exper_splsda</span>(icu_sim)</span>
<span id="cb107-6"><a href="vertical-integration.html#cb107-6" tabindex="-1"></a><span class="fu">plotIndiv</span>(fit_sim)</span></code></pre></div>
<p><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="vertical-integration.html#cb108-1" tabindex="-1"></a>fit<span class="sc">$</span>weights</span></code></pre></div>
<pre><code>##              comp1     comp2
## Bacteria 0.8580402 0.8191458
## Fungi    0.6513668 0.4845247
## Viruses  0.6118091 0.8047067</code></pre>
<p><strong>Exercise</strong>: Modify the simulator above so that the 16S group no longer depends
on disease cateogry. This will allow us to study how the integrative analysis
output changes when the data are not alignable.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="vertical-integration.html#cb110-1" tabindex="-1"></a>null_simulator <span class="ot">&lt;-</span> simulator</span>
<span id="cb110-2"><a href="vertical-integration.html#cb110-2" tabindex="-1"></a><span class="co"># fill this in</span></span>
<span id="cb110-3"><a href="vertical-integration.html#cb110-3" tabindex="-1"></a><span class="co"># null_simulator[[1]] &lt;- ???</span></span></code></pre></div>
<p><strong>Solution</strong>: We need to define a new link that no longer depends on <code>Category</code>. One solution
is to modify the existing simulator in place using <code>mutate</code>.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="vertical-integration.html#cb111-1" tabindex="-1"></a>null_simulator[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> simulator[[<span class="dv">1</span>]] <span class="sc">|&gt;</span></span>
<span id="cb111-2"><a href="vertical-integration.html#cb111-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">180</span>, <span class="at">link =</span> <span class="sc">~</span><span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb111-3"><a href="vertical-integration.html#cb111-3" tabindex="-1"></a>  <span class="fu">estimate</span>(<span class="at">nu =</span> <span class="fl">0.05</span>)</span></code></pre></div>
<p>Since we are modifying all taxa, a simpler solution is to just define a
new simulator from scratch.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="vertical-integration.html#cb112-1" tabindex="-1"></a>null_simulator[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">setup_simulator</span>(icu[[<span class="dv">1</span>]], <span class="sc">~</span><span class="dv">1</span>, <span class="sc">~</span> <span class="fu">GaussianLSS</span>()) <span class="sc">|&gt;</span></span>
<span id="cb112-2"><a href="vertical-integration.html#cb112-2" tabindex="-1"></a>  <span class="fu">estimate</span>(<span class="at">nu =</span> <span class="fl">0.05</span>)</span></code></pre></div>
</div>
<div id="simulation-results" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> Simulation Results<a href="vertical-integration.html#simulation-results" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We can rerun the integrative analysis using the modified simulator. Somewhat
surprisingly, the disease association in the bacteria group hasnât been erased.
This is an artifact of the integration. The other assays have associations with
disease group, and since the method encourages outputs across tables to be
consistent with one another, we have artificially introduced some structure into
the bacteria visualization (even if it is quite weak.) Nonetheless, we still
observe a large dropoff in weight for the bacterial table. Further, there seems
to be a minor deterioration in the group separations for the fungal and virus
communities, and the component weights are higher when we work with only the
fungal and virus assays. Altogether, this suggests that we may want to check
table-level associations with the response variable, especially if any of the
integration outputs are ambiguous. In this case, we might be able to increase
power by focusing only on class-associated assays. Nonetheless, the block
sPLS-DA also seems relatively robust â considering the dramatic change in the
microbiome table, the output for the remaining tables still surfaces interesting
relationships.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="vertical-integration.html#cb113-1" tabindex="-1"></a>icu_sim <span class="ot">&lt;-</span> <span class="fu">join_copula</span>(null_simulator, <span class="fu">copula_adaptive</span>()) <span class="sc">|&gt;</span></span>
<span id="cb113-2"><a href="vertical-integration.html#cb113-2" tabindex="-1"></a>  <span class="fu">sample</span>() <span class="sc">|&gt;</span></span>
<span id="cb113-3"><a href="vertical-integration.html#cb113-3" tabindex="-1"></a>  <span class="fu">split_assays</span>()</span>
<span id="cb113-4"><a href="vertical-integration.html#cb113-4" tabindex="-1"></a>fit_null <span class="ot">&lt;-</span> <span class="fu">exper_splsda</span>(icu_sim)</span>
<span id="cb113-5"><a href="vertical-integration.html#cb113-5" tabindex="-1"></a><span class="fu">plotIndiv</span>(fit)</span></code></pre></div>
<p><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-43-1.png" width="672" /></p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="vertical-integration.html#cb114-1" tabindex="-1"></a>fit_null<span class="sc">$</span>weights</span></code></pre></div>
<pre><code>##              comp1     comp2
## Bacteria 0.6658170 0.6571177
## Fungi    0.7479665 0.6797712
## Viruses  0.7015960 0.7929047</code></pre>
<p><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-44-1.png" width="672" /><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-44-2.png" width="672" /><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-44-3.png" width="672" /><img src="simulation_for_microbiome_files/figure-html/unnamed-chunk-44-4.png" width="672" /></p>
</div>
<div id="alignability" class="section level2 hasAnchor" number="6.5">
<h2><span class="header-section-number">6.5</span> Alignability<a href="vertical-integration.html#alignability" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>How do the canonical correlations compare between the real data and a reference
null? First, we can compute the CCA canonical correlations using the real
Bacteria and Virus data.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="vertical-integration.html#cb116-1" tabindex="-1"></a>merged <span class="ot">&lt;-</span> simulator[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)] <span class="sc">|&gt;</span></span>
<span id="cb116-2"><a href="vertical-integration.html#cb116-2" tabindex="-1"></a>  <span class="fu">join_copula</span>(<span class="fu">copula_adaptive</span>(<span class="at">thr =</span> <span class="fl">0.01</span>))</span>
<span id="cb116-3"><a href="vertical-integration.html#cb116-3" tabindex="-1"></a></span>
<span id="cb116-4"><a href="vertical-integration.html#cb116-4" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">sample</span>(merged) <span class="sc">|&gt;</span></span>
<span id="cb116-5"><a href="vertical-integration.html#cb116-5" tabindex="-1"></a>  <span class="fu">assay</span>() <span class="sc">|&gt;</span></span>
<span id="cb116-6"><a href="vertical-integration.html#cb116-6" tabindex="-1"></a>  <span class="fu">t</span>()</span>
<span id="cb116-7"><a href="vertical-integration.html#cb116-7" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">map</span>(icu, rownames)</span>
<span id="cb116-8"><a href="vertical-integration.html#cb116-8" tabindex="-1"></a>cca_result <span class="ot">&lt;-</span> <span class="fu">rcc</span>(xy[, ix[[<span class="dv">1</span>]]], xy[, ix[[<span class="dv">3</span>]]], <span class="at">method =</span> <span class="st">&quot;shrinkage&quot;</span>)</span></code></pre></div>
<p>Next, we define a reference null and compute canonical correlations on ten
replicates from this null. To define the reference null, we zero out any
correlations across tables. We then draw 100 samples from the resulting
distribution and compute canonical correlations for each.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="vertical-integration.html#cb117-1" tabindex="-1"></a>rho_null <span class="ot">&lt;-</span> <span class="fu">copula_parameters</span>(merged)</span>
<span id="cb117-2"><a href="vertical-integration.html#cb117-2" tabindex="-1"></a><span class="fu">dimnames</span>(rho_null) <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">colnames</span>(xy), <span class="fu">colnames</span>(xy))</span>
<span id="cb117-3"><a href="vertical-integration.html#cb117-3" tabindex="-1"></a>rho_null[ix[[<span class="dv">1</span>]], ix[[<span class="dv">3</span>]]] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb117-4"><a href="vertical-integration.html#cb117-4" tabindex="-1"></a>rho_null[ix[[<span class="dv">3</span>]], ix[[<span class="dv">1</span>]]] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb117-5"><a href="vertical-integration.html#cb117-5" tabindex="-1"></a></span>
<span id="cb117-6"><a href="vertical-integration.html#cb117-6" tabindex="-1"></a>null_cancor <span class="ot">&lt;-</span> merged <span class="sc">|&gt;</span></span>
<span id="cb117-7"><a href="vertical-integration.html#cb117-7" tabindex="-1"></a>  <span class="fu">mutate_correlation</span>(rho_null)</span>
<span id="cb117-8"><a href="vertical-integration.html#cb117-8" tabindex="-1"></a></span>
<span id="cb117-9"><a href="vertical-integration.html#cb117-9" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb117-10"><a href="vertical-integration.html#cb117-10" tabindex="-1"></a>cancors <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb117-11"><a href="vertical-integration.html#cb117-11" tabindex="-1"></a><span class="cf">for</span> (b <span class="cf">in</span> <span class="fu">seq_len</span>(B)) {</span>
<span id="cb117-12"><a href="vertical-integration.html#cb117-12" tabindex="-1"></a>  xy_null <span class="ot">&lt;-</span> <span class="fu">sample</span>(null_cancor) <span class="sc">|&gt;</span></span>
<span id="cb117-13"><a href="vertical-integration.html#cb117-13" tabindex="-1"></a>    <span class="fu">assay</span>() <span class="sc">|&gt;</span></span>
<span id="cb117-14"><a href="vertical-integration.html#cb117-14" tabindex="-1"></a>    <span class="fu">t</span>()</span>
<span id="cb117-15"><a href="vertical-integration.html#cb117-15" tabindex="-1"></a></span>
<span id="cb117-16"><a href="vertical-integration.html#cb117-16" tabindex="-1"></a>  cancors[[<span class="fu">glue</span>(<span class="st">&quot;sim_{b}&quot;</span>)]] <span class="ot">&lt;-</span> <span class="fu">rcc</span>(xy_null[, ix[[<span class="dv">1</span>]]], xy_null[, ix[[<span class="dv">3</span>]]], <span class="at">method =</span> <span class="st">&quot;shrinkage&quot;</span>)<span class="sc">$</span>cor</span>
<span id="cb117-17"><a href="vertical-integration.html#cb117-17" tabindex="-1"></a>}</span></code></pre></div>
<p>To help with later visualization, we combine both the real and reference
canonical correlations into a tidy data.frame. These are visualized in a hidden
chunk below â if you want to see the source code, you can review it
<a href="https://github.com/krisrs1128/microbiome-simulation/blob/main/book/04-integration.Rmd">here</a>.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="vertical-integration.html#cb118-1" tabindex="-1"></a>cancors[[<span class="st">&quot;true&quot;</span>]] <span class="ot">&lt;-</span> cca_result<span class="sc">$</span>cor</span>
<span id="cb118-2"><a href="vertical-integration.html#cb118-2" tabindex="-1"></a>contrast_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(cancors, <span class="at">.id =</span> <span class="st">&quot;rep&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb118-3"><a href="vertical-integration.html#cb118-3" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>rep, <span class="at">names_to =</span> <span class="st">&quot;loading&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb118-4"><a href="vertical-integration.html#cb118-4" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb118-5"><a href="vertical-integration.html#cb118-5" tabindex="-1"></a>    <span class="at">source =</span> <span class="fu">grepl</span>(<span class="st">&quot;true&quot;</span>, rep),</span>
<span id="cb118-6"><a href="vertical-integration.html#cb118-6" tabindex="-1"></a>    <span class="at">loading =</span> <span class="fu">as.integer</span>(loading)</span>
<span id="cb118-7"><a href="vertical-integration.html#cb118-7" tabindex="-1"></a>  )</span>
<span id="cb118-8"><a href="vertical-integration.html#cb118-8" tabindex="-1"></a><span class="fu">head</span>(contrast_data)</span></code></pre></div>
<pre><code>## # A tibble: 6 Ã 4
##   rep   loading value source
##   &lt;chr&gt;   &lt;int&gt; &lt;dbl&gt; &lt;lgl&gt; 
## 1 sim_1       1 0.913 FALSE 
## 2 sim_1       2 0.880 FALSE 
## 3 sim_1       3 0.848 FALSE 
## 4 sim_1       4 0.826 FALSE 
## 5 sim_1       5 0.815 FALSE 
## 6 sim_1       6 0.797 FALSE</code></pre>
<p><img src="simulation_for_microbiome_files/figure-html/contrast-vis-1.png" width="672" /></p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="vertical-integration.html#cb120-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.3.2 (2023-10-31)
## Platform: aarch64-apple-darwin20 (64-bit)
## Running under: macOS Sonoma 14.5
## 
## Matrix products: default
## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: Australia/Melbourne
## tzcode source: internal
## 
## attached base packages:
##  [1] splines   parallel  stats4    stats     graphics  grDevices
##  [7] utils     datasets  methods   base     
## 
## other attached packages:
##  [1] doRNG_1.8.6                    
##  [2] rngtools_1.5.2                 
##  [3] foreach_1.5.2                  
##  [4] ggpubr_0.6.0                   
##  [5] ggrepel_0.9.5                  
##  [6] edgeR_4.0.16                   
##  [7] limma_3.58.1                   
##  [8] ANCOMBC_2.4.0                  
##  [9] stringr_1.5.1                  
## [10] ks_1.14.2                      
## [11] tidySummarizedExperiment_1.12.0
## [12] ttservice_0.4.1                
## [13] SimBench_0.99.1                
## [14] wesanderson_0.3.7              
## [15] mia_1.10.0                     
## [16] MultiAssayExperiment_1.28.0    
## [17] TreeSummarizedExperiment_2.10.0
## [18] Biostrings_2.70.3              
## [19] XVector_0.42.0                 
## [20] SingleCellExperiment_1.24.0    
## [21] gamlss_5.4-22                  
## [22] nlme_3.1-165                   
## [23] gamlss.dist_6.1-1              
## [24] gamlss.data_6.0-6              
## [25] scDesigner_0.0.0.9000          
## [26] purrr_1.0.2                    
## [27] MIGsim_0.0.0.9000              
## [28] tidyr_1.3.1                    
## [29] tibble_3.2.1                   
## [30] scico_1.5.0                    
## [31] pwr_1.3-0                      
## [32] patchwork_1.2.0                
## [33] mutoss_0.1-13                  
## [34] mvtnorm_1.2-5                  
## [35] mixOmics_6.26.0                
## [36] lattice_0.22-6                 
## [37] MASS_7.3-60.0.1                
## [38] glue_1.7.0                     
## [39] ggplot2_3.5.1                  
## [40] ggdist_3.3.2                   
## [41] gamboostLSS_2.0-7              
## [42] mboost_2.9-10                  
## [43] stabs_0.6-4                    
## [44] forcats_1.0.0                  
## [45] dplyr_1.1.4                    
## [46] SummarizedExperiment_1.32.0    
## [47] Biobase_2.62.0                 
## [48] GenomicRanges_1.54.1           
## [49] GenomeInfoDb_1.38.8            
## [50] IRanges_2.36.0                 
## [51] S4Vectors_0.40.2               
## [52] BiocGenerics_0.48.1            
## [53] MatrixGenerics_1.14.0          
## [54] matrixStats_1.3.0              
## [55] SpiecEasi_1.1.3                
## [56] CovTools_0.5.4                 
## 
## loaded via a namespace (and not attached):
##   [1] igraph_2.0.3                ica_1.0-3                  
##   [3] plotly_4.10.4               Formula_1.2-5              
##   [5] scater_1.30.1               BiocBaseUtils_1.4.0        
##   [7] zlibbioc_1.48.2             tidyselect_1.2.1           
##   [9] bit_4.0.5                   doParallel_1.0.17          
##  [11] pspline_1.0-20              blob_1.2.4                 
##  [13] S4Arrays_1.2.1              png_0.1-8                  
##  [15] plotrix_3.8-4               cli_3.6.3                  
##  [17] CVXR_1.0-14                 pulsar_0.3.11              
##  [19] arrayhelpers_1.1-0          multtest_2.58.0            
##  [21] goftest_1.2-3               textshaping_0.4.0          
##  [23] bluster_1.12.0              BiocNeighbors_1.20.2       
##  [25] uwot_0.2.2                  mime_0.12                  
##  [27] evaluate_0.24.0             tidytree_0.4.6             
##  [29] leiden_0.4.3.1              stringi_1.8.4              
##  [31] backports_1.5.0             XML_3.99-0.17              
##  [33] lmerTest_3.1-3              gsl_2.1-8                  
##  [35] Exact_3.2                   httpuv_1.6.15              
##  [37] AnnotationDbi_1.64.1        magrittr_2.0.3             
##  [39] mclust_6.1.1                ADGofTest_0.3              
##  [41] pcaPP_2.0-4                 sctransform_0.4.1          
##  [43] rootSolve_1.8.2.4           ggbeeswarm_0.7.2           
##  [45] DBI_1.2.3                   genefilter_1.84.0          
##  [47] jquerylib_0.1.4             withr_3.0.0                
##  [49] corpcor_1.6.10              systemfonts_1.1.0          
##  [51] class_7.3-22                lmtest_0.9-40              
##  [53] sva_3.50.0                  compositions_2.0-8         
##  [55] htmlwidgets_1.6.4           fs_1.6.4                   
##  [57] labeling_0.4.3              SparseArray_1.2.4          
##  [59] DEoptimR_1.1-3              DESeq2_1.42.1              
##  [61] cellranger_1.1.0            tidybayes_3.0.6            
##  [63] extrafont_0.19              annotate_1.80.0            
##  [65] flare_1.7.0.1               lmom_3.0                   
##  [67] reticulate_1.37.0           minpack.lm_1.2-4           
##  [69] geigen_2.3                  zoo_1.8-12                 
##  [71] knitr_1.48                  nnls_1.5                   
##  [73] svUnit_1.0.6                SHT_0.1.8                  
##  [75] decontam_1.22.0             fansi_1.0.6                
##  [77] grid_4.3.2                  rhdf5_2.46.1               
##  [79] data.table_1.15.4           ruv_0.9.7.1                
##  [81] vegan_2.6-6.1               RSpectra_0.16-1            
##  [83] irlba_2.3.5.1               extrafontdb_1.0            
##  [85] fastDummies_1.7.3           DescTools_0.99.54          
##  [87] ellipsis_0.3.2              ade4_1.7-22                
##  [89] lazyeval_0.2.2              yaml_2.3.9                 
##  [91] survival_3.7-0              scattermore_1.2            
##  [93] crayon_1.5.3                tensorA_0.36.2.1           
##  [95] phyloseq_1.46.0             RcppAnnoy_0.0.22           
##  [97] RColorBrewer_1.1-3          progressr_0.14.0           
##  [99] later_1.3.2                 ggridges_0.5.6             
## [101] codetools_0.2-20            base64enc_0.1-3            
## [103] Seurat_5.1.0                KEGGREST_1.42.0            
## [105] Rtsne_0.17                  shape_1.4.6.1              
## [107] randtoolbox_2.0.4           foreign_0.8-87             
## [109] pkgconfig_2.0.3             scatterplot3d_0.3-44       
## [111] spatstat.sparse_3.1-0       ape_5.8                    
## [113] viridisLite_0.4.2           xtable_1.8-4               
## [115] highr_0.11                  car_3.1-2                  
## [117] fastcluster_1.2.6           plyr_1.8.9                 
## [119] httr_1.4.7                  rbibutils_2.2.16           
## [121] tools_4.3.2                 globals_0.16.3             
## [123] SeuratObject_5.0.2          kde1d_1.0.7                
## [125] beeswarm_0.4.0              htmlTable_2.4.2            
## [127] broom_1.0.6                 checkmate_2.3.1            
## [129] lme4_1.1-35.5               rgl_1.3.1                  
## [131] assertthat_0.2.1            rngWELL_0.10-9             
## [133] digest_0.6.36               bayesm_3.1-6               
## [135] permute_0.9-7               numDeriv_2016.8-1.1        
## [137] bookdown_0.40               Matrix_1.6-5               
## [139] farver_2.1.2                reshape2_1.4.4             
## [141] yulab.utils_0.1.4           viridis_0.6.5              
## [143] DirichletMultinomial_1.44.0 rpart_4.1.23               
## [145] cachem_1.1.0                polyclip_1.10-6            
## [147] WGCNA_1.72-5                Hmisc_5.1-3                
## [149] generics_0.1.3              dynamicTreeCut_1.63-1      
## [151] parallelly_1.37.1           biomformat_1.30.0          
## [153] statmod_1.5.0               impute_1.76.0              
## [155] huge_1.3.5                  ragg_1.3.2                 
## [157] RcppHNSW_0.6.0              ScaledMatrix_1.10.0        
## [159] minqa_1.2.7                 carData_3.0-5              
## [161] pbapply_1.7-2               glmnet_4.1-8               
## [163] spam_2.10-0                 utf8_1.2.4                 
## [165] gtools_3.9.5                shapes_1.2.7               
## [167] readxl_1.4.3                preprocessCore_1.64.0      
## [169] inum_1.0-5                  ggsignif_0.6.4             
## [171] energy_1.7-11               gridExtra_2.3              
## [173] shiny_1.8.1.1               GenomeInfoDbData_1.2.11    
## [175] rhdf5filters_1.14.1         RCurl_1.98-1.16            
## [177] memoise_2.0.1               rmarkdown_2.27             
## [179] scales_1.3.0                stabledist_0.7-1           
## [181] partykit_1.2-20             gld_2.6.6                  
## [183] svglite_2.1.3               future_1.33.2              
## [185] RANN_2.6.1                  spatstat.data_3.1-2        
## [187] rstudioapi_0.16.0           cluster_2.1.6              
## [189] spatstat.utils_3.0-5        hms_1.1.3                  
## [191] fitdistrplus_1.1-11         munsell_0.5.1              
## [193] cowplot_1.1.3               colorspace_2.1-0           
## [195] ellipse_0.5.0               rlang_1.1.4                
## [197] quadprog_1.5-8              copula_1.1-3               
## [199] DelayedMatrixStats_1.24.0   sparseMatrixStats_1.14.0   
## [201] dotCall64_1.1-1             scuttle_1.12.0             
## [203] mgcv_1.9-1                  xfun_0.45                  
## [205] coda_0.19-4.1               e1071_1.7-14               
## [207] TH.data_1.1-2               posterior_1.5.0            
## [209] iterators_1.0.14            rARPACK_0.11-0             
## [211] abind_1.4-5                 libcoin_1.0-10             
## [213] treeio_1.26.0               gmp_0.7-4                  
## [215] Rhdf5lib_1.24.2             DECIPHER_2.30.0            
## [217] bitops_1.0-7                Rdpack_2.6                 
## [219] promises_1.3.0              RSQLite_2.3.7              
## [221] sandwich_3.1-0              DelayedArray_0.28.0        
## [223] proxy_0.4-27                Rmpfr_0.9-5                
## [225] GO.db_3.18.0                compiler_4.3.2             
## [227] prettyunits_1.2.0           boot_1.3-30                
## [229] distributional_0.4.0        beachmat_2.18.1            
## [231] listenv_0.9.1               Rcpp_1.0.12                
## [233] Rttf2pt1_1.3.12             BiocSingular_1.18.0        
## [235] tensor_1.5                  progress_1.2.3             
## [237] BiocParallel_1.36.0         insight_0.20.1             
## [239] spatstat.random_3.2-3       R6_2.5.1                   
## [241] fastmap_1.2.0               multcomp_1.4-25            
## [243] rstatix_0.7.2               vipor_0.4.7                
## [245] ROCR_1.0-11                 rsvd_1.0.5                 
## [247] nnet_7.3-19                 gtable_0.3.5               
## [249] KernSmooth_2.23-24          miniUI_0.1.1.1             
## [251] deldir_2.0-4                htmltools_0.5.8.1          
## [253] ggthemes_5.1.0              bit64_4.0.5                
## [255] spatstat.explore_3.2-7      lifecycle_1.0.4            
## [257] nloptr_2.1.1                sass_0.4.9                 
## [259] vctrs_0.6.5                 robustbase_0.99-3          
## [261] VGAM_1.1-11                 spatstat.geom_3.2-9        
## [263] sp_2.1-4                    future.apply_1.11.2        
## [265] pracma_2.4.4                rvinecopulib_0.6.3.1.1     
## [267] bslib_0.7.0                 pillar_1.9.0               
## [269] locfit_1.5-9.10             jsonlite_1.8.8             
## [271] expm_0.999-9</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="batch-effect-correction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/05-vertical-integration.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["simulation_for_microbiome.pdf", "simulation_for_microbiome.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
